File: v4/rust/iprange-livedb/src/writer_core/reclaim.rs
//! Bounded reclamation draft preparation.

use crate::cancellation::CancellationToken;
use crate::draft_store::{Draft, DraftStore};
use crate::error::Result;
use crate::random;

use super::{CommitAttempt, WriterCore};

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub(crate) struct PreparedReclamation {
    pub(crate) transaction_count: u64,
    pub(crate) page_count: u64,
    pub(crate) attempt: CommitAttempt,
}

impl WriterCore {
    pub(crate) fn prepare_reclamation(
        &mut self,
        oldest_reader: Option<u64>,
        max_transactions: u64,
        max_pages: u64,
        cancellation: &CancellationToken,
    ) -> Result<Option<PreparedReclamation>> {
        let mut draft = Draft::new(self.base.meta, random::nonzero_128()?)?;
        let mut checkpoint = || cancellation.check();
        let selection = DraftStore::new(
            &mut self.mapping,
            self.base.meta.page_count,
            self.budget,
            &mut draft,
        )
        .select_reclamation(oldest_reader, max_transactions, max_pages, &mut checkpoint)?;
        let Some(selection) = selection else {
            return Ok(None);
        };

        let attempt = CommitAttempt {
            database_id: draft.meta.database_id,
            transaction_id: draft.meta.txn_id,
            commit_nonce: draft.meta.commit_nonce,
        };
        self.draft = Some(draft);
        let transaction_count = selection.transactions;
        let page_count = selection.pages;
        DraftStore::new(
            &mut self.mapping,
            self.base.meta.page_count,
            self.budget,
            self.draft
                .as_mut()
                .expect("reclamation draft was installed"),
        )
        .apply_reclamation(selection, &mut checkpoint)?;
        self.prepare(cancellation)?;
        cancellation.check()?;
        Ok(Some(PreparedReclamation {
            transaction_count,
            page_count,
            attempt,
        }))
    }
}

File: v4/go/internal/writer/reclaim.go
// Bounded reclamation draft preparation (Rust writer_core/reclaim.rs).
// Reclamation selects the oldest reader-safe retired transactions within
// the work limits, installs a reclamation draft, applies the selection
// physically, and prepares it for publication exactly like a normal
// commit.

package writer

import (
	"crypto/rand"

	"github.com/firehol/iprange/v4/go/internal/format"
)

// PreparedReclamation is the installed, prepared reclamation draft (Rust
// writer_core::PreparedReclamation).
type PreparedReclamation struct {
	TransactionCount uint64
	PageCount        uint64
	Attempt          CommitAttempt
}

// PrepareReclamation selects bounded reclamation work and leaves a
// prepared draft ready for Publish (Rust WriterCore::prepare_reclamation).
// oldestReader, when non-nil, is the oldest active reader transaction;
// extents retired by that transaction or later are never reclaimed.
func (c *Core) PrepareReclamation(oldestReader *uint64, maxTransactions, maxPages uint64, checkpoint func() error) (*PreparedReclamation, error) {
	if err := c.requireHealthy(); err != nil {
		return nil, err
	}
	nonce, err := randomNonce()
	if err != nil {
		return nil, err
	}
	draft, err := NewDraft(c.base.Meta, nonce)
	if err != nil {
		return nil, err
	}
	store := NewDraftStore(c.m, c.base.Meta.PageCount, c.budget, draft)
	selection, err := store.SelectReclamation(oldestReader, maxTransactions, maxPages, checkpoint)
	if err != nil {
		return nil, err
	}
	if selection == nil {
		return nil, nil
	}
	attempt := CommitAttempt{
		DatabaseID:    draft.meta.DatabaseID,
		TransactionID: draft.meta.TxnID,
		CommitNonce:   draft.meta.CommitNonce,
	}
	c.draft = draft
	transactionCount := selection.Transactions
	pageCount := selection.Pages
	if err := store.ApplyReclamation(selection, checkpoint); err != nil {
		return nil, err
	}
	if err := c.Prepare(checkpoint); err != nil {
		return nil, err
	}
	if checkpoint != nil {
		if err := checkpoint(); err != nil {
			return nil, err
		}
	}
	return &PreparedReclamation{
		TransactionCount: transactionCount,
		PageCount:        pageCount,
		Attempt:          attempt,
	}, nil
}

// randomNonce draws one nonzero 128-bit commit nonce (Rust
// random::nonzero_128: one fill, all-zero is a hard error).
func randomNonce() ([16]byte, error) {
	var nonce [16]byte
	if _, err := rand.Read(nonce[:]); err != nil {
		return nonce, &format.Error{Code: format.CodeIO, Detail: "commit nonce: " + err.Error()}
	}
	if nonce == [16]byte{} {
		return nonce, &format.Error{Code: format.CodeFormatInvalid, Detail: "operating-system randomness returned an all-zero identity"}
	}
	return nonce, nil
}
